channeldb/addr_source_test.go

```diff
@@ -0,0 +1,109 @@
+package channeldb
+
+import (
+	"net"
+	"testing"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+var (
+	addr1 = &net.TCPAddr{IP: (net.IP)([]byte{0x1}), Port: 1}
+	addr2 = &net.TCPAddr{IP: (net.IP)([]byte{0x2}), Port: 2}
+	addr3 = &net.TCPAddr{IP: (net.IP)([]byte{0x3}), Port: 3}
+)
+
+// TestMultiAddrSource tests that the multiAddrSource correctly merges and
+// deduplicates the results of a set of AddrSource implementations.
+func TestMultiAddrSource(t *testing.T) {
+	t.Parallel()
+
+	var pk1 = newTestPubKey(t)
+
+	t.Run("both sources have results", func(t *testing.T) {
+		t.Parallel()
+
+		var (
+			src1 = newMockAddrSource(t)
+			src2 = newMockAddrSource(t)
+		)
+
+		// Let source 1 know of 2 addresses (addr 1 and 2) for node 1.
+		src1.On("AddrsForNode", pk1).Return(
+			[]net.Addr{addr1, addr2}, nil,
+		)
+
+		// Let source 2 know of 2 addresses (addr 2 and 3) for node 1.
+		src2.On("AddrsForNode", pk1).Return(
+			[]net.Addr{addr2, addr3}, nil,
+		)
+
+		// Create a multi-addr source that consists of both source 1
+		// and 2.
+		multiSrc := NewMultiAddrSource(src1, src2)
+
+		// Query it for the addresses known for node 1. The results
+		// should contain addr 1, 2 and 3.
+		addrs, err := multiSrc.AddrsForNode(pk1)
+		require.NoError(t, err)
+		require.ElementsMatch(t, addrs, []net.Addr{addr1, addr2, addr3})
+	})
+
+	t.Run("only once source has results", func(t *testing.T) {
+		t.Parallel()
+
+		var (
+			src1 = newMockAddrSource(t)
+			src2 = newMockAddrSource(t)
+		)
+
+		// Let source 1 know of address 1 for node 1.
+		src1.On("AddrsForNode", pk1).Return([]net.Addr{addr1}, nil)
+		src2.On("AddrsForNode", pk1).Return(nil, nil)
+
+		// Create a multi-addr source that consists of both source 1
+		// and 2.
+		multiSrc := NewMultiAddrSource(src1, src2)
+
+		// Query it for the addresses known for node 1. The results
+		// should contain addr 1.
+		addrs, err := multiSrc.AddrsForNode(pk1)
+		require.NoError(t, err)
+		require.ElementsMatch(t, addrs, []net.Addr{addr1})
+	})
+
+}
+
+type mockAddrSource struct {
+	t *testing.T
+	mock.Mock
+}
+
+var _ AddrSource = (*mockAddrSource)(nil)
+
+func newMockAddrSource(t *testing.T) *mockAddrSource {
+	return &mockAddrSource{t: t}
+}
+
+func (m *mockAddrSource) AddrsForNode(pub *btcec.PublicKey) ([]net.Addr,
+	error) {
+
+	args := m.Called(pub)
+	if args.Get(0) == nil {
+		return nil, args.Error(1)
+	}
+
+	addrs, ok := args.Get(0).([]net.Addr)
+	require.True(m.t, ok)
```

Comment:
Could we just return `args.Get(0).([]net.Addr)`?

Reply:
tried this but the linter didnt like it

Reply:
i see - think we ignore the `forcetypeassert` for all the mock files, sth like `mocks.go`